judge/views/user.py

```diff
@@ -507,3 +512,98 @@ def post(self, request, *args, **kwargs):
+
+    def form_valid(self, form):
+        signer = signing.TimestampSigner()
+        new_email = form.cleaned_data['email']
+        activation_key = base64.urlsafe_b64encode(signer.sign_object({
+            'id': self.request.user.id,
+            'email': new_email,
+        }).encode()).decode()
+
+        current_site = get_current_site(self.request)
+        context = {
+            'domain': current_site.domain,
+            'site_name': current_site.name,
+            'protocol': 'https' if self.request.is_secure() else 'http',
+            'site_admin_email': settings.SITE_ADMIN_EMAIL,
+            'expiry_minutes': settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES,
+            'user': self.request.user,
+            'activation_key': activation_key,
+            'new_email': new_email,
+        }
+        send_mail(
+            self.notify_subject_template_name, self.notify_email_template_name, context, None, self.request.user.email,
+            self.notify_html_email_template_name,
+        )
+        send_mail(
+            self.activate_subject_template_name, self.activate_email_template_name, context, None, new_email,
+            self.activate_html_email_template_name,
+        )
+
+        return generic_message(
+            self.request,
+            _('Email change requested'),
+            _('Please click on the link sent to the new email address.'),
+        )
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
+        return kwargs
+
+    def post(self, request, *args, **kwargs):
+        key = f'emailchange!{request.META["REMOTE_ADDR"]}'
+        cache.add(key, 0, timeout=settings.DMOJ_EMAIL_CHANGE_LIMIT_WINDOW)
+        if cache.incr(key) > settings.DMOJ_EMAIL_CHANGE_LIMIT_COUNT:
+            return HttpResponse(_('You have sent too many email change requests. Please try again later.'),
+                                content_type='text/plain', status=429)
+        return super().post(request, *args, **kwargs)
+
+
+class EmailChangeActivateView(LoginRequiredMixin, View):
+    def get(self, request, *args, **kwargs):
+        activation_key = kwargs['activation_key']
+        signer = signing.TimestampSigner()
+        try:
+            try:
+                data = signer.unsign_object(
+                    base64.urlsafe_b64decode(activation_key.encode()).decode(),
+                    max_age=settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES * 60,
+                )
+            except (binascii.Error, signing.BadSignature):
+                raise ValueError(_('Invalid activation key. Please try again.'))
+            except signing.SignatureExpired:
+                raise ValueError(_('This request is expired. Please try again.'))
+            if data['id'] != request.user.id:
+                raise ValueError(
+                    _('Please try again from the account this email change was originally requested from.'),
```

Comment:
I feel like an email change confirmation shouldn't require a login. I often click on confirmation links on my phone that i requested on my computer.

Reply:
Hmm, I'm open to discussion on this. I do think there's some merit to ensuring the user is logged in before switching emails. This prevents a user mistyping an email that is actually in use, and then the user of the mistyped email accidentally clicking on the link causing the email to be switched.

Reply:
Like you mean user A requests a change to email B, and the owner of B can now take over the account of A...

Ok, granted, that seems bad. I think this is fine then.